pep-9999.rst

```diff
@@ -0,0 +1,464 @@
+multicore performance in Python.
+
+This work will take place in the following discrete steps:
+
+1. Atomic reference counting of immutable objects to permit 
+   concurrent increments and decrements on shared object RC's.
+2. Support for managing cyclic immutable data using reference
+   counting.
+3. Support for sharing immutable data across subinterpreters. 
+4. Support for sharing mutable data across subinterpreters,
+   with dynamic ownership protecting against data races.
+5. Support for behaviour-oriented concurrency.
+
+
+Support for Atomic Reference Counting
+-------------------------------------
+
+As preparation for the extension `Sharing Immutable Data Across Subinterpreters`_,
+we will add support for atomic reference counting for immutable objects. This
+will complement work in `Simplified Garbage Collection for Immutable Object Graphs`_,
+which aims to make memory management of immutable data more efficient.
+
+When immutable data is shared across threads we must ensure that 
+concurrent reference count manipulations are correct, which in turns
+requires atomic increments and decrements. Note that since we are only
+planning to share immutable objects across different GIL's, it is
+*not* possible for two threads to read--write or write--write race
+on a single field. Thus we only need to protect the reference counter
+manipulations, avoiding most of the complexity of `PEP 703`_
+
+
+Simplified Garbage Collection for Immutable Object Graphs
+---------------------------------------------------------
+
+In `previous work <https://dl.acm.org/doi/10.1145/3652024.3665507>`_,
+we have identified that objects that make up cyclic immutable
+garbage will always have the same lifetime. This means that a
+single reference count could be used to track the lifetimes of
+all the objects in such a strongly connected component (SCC).
+
+We plan to extend the freeze function with a SCC analysis that
+creates a designated (atomic) reference count for the entire
+SCC, such that reference count manipulations on any object in
+the SCC will be "forwarded" to that shared reference count.
+This can be done without bloating objects by repurposing the
+existing reference counter data to be used as a pointer to
+the shared counter.
+
+This technique permits handling cyclic garbage using plain
+reference counting, and because of the single reference count
+for an entire SCC, we will detect when all the objects in the
+SCC expire at once.
+
+
+Sharing Immutable Data Across Subinterpreters
+---------------------------------------------
+
+We plan to extend the functionality of `multiple subinterpreters <https://peps.python.org/pep-0734/>`_
+to *share* immutable data without copying. This is safe and
+efficient as it avoids the copying or serialisation when
+objects are transmitted across subinterpreters.
+
+This change will require reference counts to be atomic (as
+discussed above) and the subclass list of a type object to
+be made thread-safe.
```

Comment:
It also requires modules loaded in one subinterpreter to be accessible from another. Implementation details here to be discussed, but the version we have been working on is one in which a side effect of calling `freezeglobals` is to make it such that all subsequent modules that are imported are imported into the main module and immediately frozen (i.e. *not* the subinterpreter) and shared across all subinterpreters.

Reply:
Integrated this text. Please take a look.